crates/ol/mempool/src/state.rs

```diff
@@ -0,0 +1,1933 @@
+            txid,
+            ssz::Encode::as_ssz_bytes(&entry.tx),
+            self.current_tip.slot(),
+            insertion_id,
+        );
+        ctx.storage.mempool().put_tx(tx_data)?;
+
+        // Add to ordering index
+        self.ordering_index.insert(priority, txid);
+
+        // Add to per-account index for cascade removal
+        let account_id = entry.tx.target();
+        self.by_account.entry(account_id).or_default().insert(txid);
+
+        // Add to entries
+        self.entries.insert(txid, entry);
+
+        // Update stats
+        self.stats.mempool_size += 1;
+        self.stats.total_bytes += tx_size;
+        self.stats.enqueues_accepted += 1;
+
+        Ok(txid)
+    }
+
+    /// Get all transactions in priority order, filtering out transactions with gaps.
+    ///
+    /// For SnarkAccountUpdate transactions, only returns transactions that are ready
+    /// (no gaps in sequence numbers before them). GenericAccountMessage transactions
+    /// are always included (no sequence number requirements).
+    ///
+    /// Returns (txid, transaction) pairs in priority order.
+    ///
+    /// Used by `handle_best_transactions()` for block assembly.
+    pub(crate) async fn get_all_transactions(
+        &mut self,
+        ctx: &MempoolContext,
+    ) -> OLMempoolResult<Vec<(OLTxId, OLMempoolTransaction)>> {
+        // Get state accessor (cached or fetched)
+        let state_accessor = self.get_or_fetch_state_accessor(ctx).await?;
+
+        // Step 1: Group transactions by account and collect seq_nos
+        // For SnarkAccountUpdate: group by account, collect seq_nos
+        // For GenericAccountMessage: always include (no seq_no requirements)
+        let mut account_seq_nos: HashMap<AccountId, HashSet<u64>> = HashMap::new();
+        let mut snark_txs: Vec<(AccountId, u64, OLTxId, OLMempoolTransaction)> = Vec::new();
+        let mut gam_txs: Vec<(OLTxId, OLMempoolTransaction)> = Vec::new();
+
+        for txid in self.ordering_index.values() {
+            if let Some(entry) = self.entries.get(txid) {
+                match entry.tx.payload() {
+                    OLMempoolTxPayload::SnarkAccountUpdate(payload) => {
+                        let account = *payload.target();
+                        let seq_no = payload.base_update.operation().seq_no();
+                        account_seq_nos.entry(account).or_default().insert(seq_no);
+                        snark_txs.push((account, seq_no, *txid, entry.tx.clone()));
+                    }
+                    OLMempoolTxPayload::GenericAccountMessage(_) => {
+                        // GenericAccountMessage transactions have no seq_no requirements
+                        gam_txs.push((*txid, entry.tx.clone()));
+                    }
+                }
+            }
+        }
+
+        // Step 2: For each account with SnarkAccountUpdate transactions, find the first gap
+        // and only include transactions before the first gap
```

Comment:
This seems redundant. These have already been filtered upon submission. We should be guaranteed not to have gaps, and block assembly will validate it anyway.